graphql/src/execution/ast.rs

```diff
@@ -0,0 +1,349 @@
+use std::{collections::HashSet, ops::Deref};
+
+use graph::{
+    components::store::EntityType,
+    data::graphql::{DocumentExt, ObjectOrInterface},
+    prelude::{anyhow, q, r, s, QueryExecutionError, Schema, ValueMap},
+};
+use graphql_parser::Pos;
+
+use crate::schema::ast::ObjectType;
+
+/// A selection set is a table that maps object types to the fields that
+/// should be selected for objects of that type. The types are always
+/// concrete object types, never interface or union types. When a
+/// `SelectionSet` is constructed, fragments must already have been resolved
+/// as it only allows using fields.
+///
+/// The set of types that a `SelectionSet` can accommodate must be set at
+/// the time the `SelectionSet` is constructed. It is not possible to add
+/// more types to it, but it is possible to add fields for all known types
+/// or only some of them
+#[derive(Debug, Clone, PartialEq)]
+pub struct SelectionSet {
+    // Map object types to the list of fields that should be selected for
+    // them
+    items: Vec<(String, Vec<Field>)>,
+}
+
+impl SelectionSet {
+    /// Create a new `SelectionSet` that can handle the given types
+    pub fn new(types: Vec<String>) -> Self {
+        let items = types.into_iter().map(|name| (name, Vec::new())).collect();
+        SelectionSet { items }
+    }
+
+    /// Create a new `SelectionSet` that can handle the same types as
+    /// `other`, but ignore all fields from `other`
+    pub fn empty_from(other: &SelectionSet) -> Self {
+        let items = other
+            .items
+            .iter()
+            .map(|(name, _)| (name.clone(), Vec::new()))
+            .collect();
+        SelectionSet { items }
+    }
+
+    /// Return `true` if this selection set does not select any fields for
+    /// its types
+    pub fn is_empty(&self) -> bool {
+        self.items.iter().all(|(_, fields)| fields.is_empty())
+    }
+
+    /// If the selection set contains a single field across all its types,
+    /// return it. Otherwise, return `None`
+    pub fn single_field(&self) -> Option<&Field> {
```

Comment:
Looking at the one place where this is used, this definition seems broader than necessary, I think it could be simplified to check for a single type with a single field.

Reply:
The difficulty is that if the selection is for an interface, the `SelectionSet` will contain entries for multiple types. In that case, `single_field` checks that they are all the same.